### PR TITLE
[ET-VK] Set up CMake build for Vulkan Delegate

### PR DESCRIPTION
**Pull request by SS-JIA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #2179

## Context

This PR sets up the CMake build for the Vulkan delegate. It adds the following files:

* `executorch/backends/vulkan/cmake/ATenVulkan.cmake`, which defines the following libraries that uses source files from the `pytorch` third party repository:
    * `vulkan_api_lib`: compiles everything under `pytorch/aten/src/ATen/native/vulkan/api`
    * `vulkan_shader_lib`: contains compiled SPIR-V binaries as well as everything under `pytorch/aten/src/ATen/native/vulkan/impl`
* `executorch/backends/vulkan/cmake/ATenVulkan.cmake`, which defines the following libraries:
  * `vulkan_graph_runtime`: compiles everything under `executorch/backends/vulkan/runtime/graph`
  * `vulkan_schema`: compiles the generated headers from `executorch/backends/vulkan/runtime/serialization/schema.fbs`
  * `vulkan_backend`: compiles C++ files under `executorch/backends/vulkan/runtime`
  * (Optional) `vulkan_compute_api_test`: executable compiled from `executorch/backends/vulkan/test/vulkan_compute_api_test.cpp`

The following options have been added to the root `CMakeLists.txt` to support building the Vulkan delegate:

* `EXECUTORCH_BUILD_VULKAN`: triggers building of all aforementioned targets
* `EXECUTORCH_BUILD_VULKAN_TESTS`: triggers building of `vulkan_compute_api_test`

I verified that the added targets can build on a linux machine as well as using the Android toolchain (NDK r20). `vulkan_compute_api_test` was able to run and pass tests successfully on both linux and Android.

Differential Revision: [D54390748](https://our.internmc.facebook.com/intern/diff/D54390748)

### CODE REVIEW

**Comment by SS-JIA** [comment]
@SS-JIA has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54390748).

**Review by larryliu0820 — APPROVED** [review]

File: backends/vulkan/cmake/ATenVulkan.cmake [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+if(NOT VULKAN_THIRD_PARTY_PATH)
+  set(VULKAN_THIRD_PARTY_PATH ${CMAKE_CURRENT_SOURCE_DIR}/../third-party)
+endif()
+
+# Shader Codegen
+
+# Trigger Shader code generation
+set(USE_VULKAN ON)
+include(${PYTORCH_PATH}/cmake/VulkanCodegen.cmake)
```
[the remark this answers is not in the document]
Comment (SS-JIA) [reply backends/vulkan/cmake/ATenVulkan.cmake:19]: good point!

File: CMakeLists.txt [hunk 2]
```diff
@@ -315,6 +315,15 @@ if(EXECUTORCH_BUILD_EXECUTOR_RUNNER)
   target_compile_options(executor_runner PUBLIC ${_common_compile_options})
 endif()
 
+# Options to build test targets
+option(EXECUTORCH_BUILD_VULKAN_TESTS
+       "Build the backends/vulkan/tests directory" OFF)
+
+# Add googletest if any test targets should be build
+if(EXECUTORCH_BUILD_VULKAN_TESTS)
```
[the remark this answers is not in the document]
Comment (SS-JIA) [reply CMakeLists.txt:323]: My initial thought was that every backend/module would have a flag that could control whether the tests are built for that backend. Like `EXECUTORCH_BUILD_XNNPACK_TESTS` for example. That way you can have very fine grain control over what is built, but thinking more about it that's too many flags to deal with lol. Will update to be more generic as suggested.

**Comment by SS-JIA** [comment]
@SS-JIA has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54390748).

**Comment by SS-JIA** [comment]
@SS-JIA has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54390748).

**Pull request closed without merging**